src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs

```diff
@@ -865,6 +876,70 @@ async Task<IEnumerable<IAccount>> ITokenCacheInternal.GetAccountsAsync(Authentic
             return accounts;
         }
 
+        async Task<IDictionary<string, ITenantProfile>> ITokenCacheInternal.GetTenantProfilesAsync(AuthenticationRequestParameters requestParameters, string homeAccountId)
+        {
+            if (requestParameters.AuthorityInfo.AuthorityType != AuthorityType.Aad || string.IsNullOrEmpty(homeAccountId))
+            {
+                return null;
+            }
+
+            IEnumerable<MsalAccountCacheItem> accountCacheItems = _accessor.GetAllAccounts();
+
+            // Multi-cloud support - must filter by env.
+            ISet<string> allEnvironmentsInCache = new HashSet<string>(
+                accountCacheItems.Select(aci => aci.Environment),
+                StringComparer.OrdinalIgnoreCase);
+
+            AdalUsersForMsal adalUsersResult = null;
+
+            if (ServiceBundle.Config.LegacyCacheCompatibilityEnabled)
+            {
+                adalUsersResult = CacheFallbackOperations.GetAllAdalUsersForMsal(
+                    Logger,
+                    LegacyCachePersistence,
+                    ClientId);
+                allEnvironmentsInCache.UnionWith(adalUsersResult.GetAdalUserEnviroments());
+            }
+
+            InstanceDiscoveryMetadataEntry instanceMetadata = await ServiceBundle.InstanceDiscoveryManager.GetMetadataEntryTryAvoidNetworkAsync(
+                requestParameters.AuthorityInfo,
+                allEnvironmentsInCache,
+                requestParameters.RequestContext).ConfigureAwait(false);
+
+            return GetTenantProfiles(instanceMetadata, homeAccountId);
+        }
+
+        private IDictionary<string, ITenantProfile> GetTenantProfiles(InstanceDiscoveryMetadataEntry instanceMetadata, string homeAccountId)
+        {
+            if (string.IsNullOrEmpty(homeAccountId) || string.IsNullOrEmpty(AccountId.ParseFromString(homeAccountId).TenantId))
+            {
+                return null;
+            }
+
+            IDictionary<string, ITenantProfile> tenantProfiles = new Dictionary<string, ITenantProfile>();
+            bool filterByClientId = !_featureFlags.IsFociEnabled;
+
+            // Get id tokens in the cache and filter them with environment and home account id to get the id tokens for the account.
+            IEnumerable<MsalIdTokenCacheItem> idTokenCacheItems = GetAllIdTokensWithNoLocks(filterByClientId);
```

Comment:
i think we are already doing this logic elsewhere in the this class, so maybe just add the part about returning the tenant profile? cache look ups are expensive.